---
title: "R Notebook"
output: html_notebook
---

# General Setup
Setup chunk

```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 8)
knitr::opts_knit$set(root.dir = normalizePath(".."))
knitr::opts_knit$get("root.dir")
```

Setup reticulate

```{r}
library(reticulate)
knitr::knit_engines$set(python = reticulate::eng_python)
py_available(initialize = FALSE)
use_python(Sys.which("python"))
py_config()
```

Load libraries

```{r}
library(Seurat)
library(simspec)
library(harmony)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(uwot)
```



# Data
Load datasets

```{r}
lizard_all = readRDS("data/expression/lizard_all_v3.RDS")
turtle_all = readRDS("data/expression/turtle_all_v3.RDS")
axolotl_all = readRDS(file = "data/expression/axolotl/palliumres07_harmony.RDS")
drerio_brain = readRDS("data/expression/drerio_brain_v3.RDS")
finches_brain = readRDS("data/expression/HVC_RA_X.RDS")
finches_list = SplitObject(finches_brain, split.by = "species")
mouse_brain = readRDS("data/expression/l5_all_seurat.RDS")
human_10x = readRDS("data/expression/human_10x.RDS")
```

Load metadatas

```{r}
lizard_meta = read.csv("data/annotations/lizard_all_umeta.csv", header = T, row.names = 1)
turtle_meta = read.csv("data/annotations/turtle_all_umeta.csv", header = T, row.names = 1)
axolotl_meta = read.csv("data/annotations/axolotl_all_umeta.csv", header = T, row.names = 1)
human_meta = read.csv("data/annotations/human10x_all_umeta.csv", header = T, row.names = 1)
mouse_meta = read.csv("data/annotations/mouse_all_umeta.csv", header = T, row.names = 1)
zebrafish_meta = read.csv("data/annotations/drerio_all_umeta.csv", header = T, row.names = 1)
bfinch_meta = read.csv("data/annotations/bfinch_all_umeta.csv", header = T, row.names = 1)
zfinch_meta = read.csv("data/annotations/zfinch_all_umeta.csv", header = T, row.names = 1)
```

Add metadata

```{r}
lizard_all = AddMetaData(lizard_all, metadata = lizard_meta)
turtle_all = AddMetaData(turtle_all, metadata = turtle_meta)
axolotl_all = AddMetaData(axolotl_all, metadata = axolotl_meta)
drerio_all = AddMetaData(drerio_brain, metadata = zebrafish_meta)
zfinch_all = AddMetaData(finches_list$ZF, metadata = zfinch_meta)
bfinch_all = AddMetaData(finches_list$BF, metadata = bfinch_meta)
mouse_all = AddMetaData(mouse_brain, metadata = mouse_meta)
human_all = AddMetaData(human_10x, metadata = human_meta)
```

Filter data (doublets and odd cells)

```{r}
# lizard_all
# human_all
# zfinch_all
turtle_all = turtle_all[,turtle_all$classes!="doublets"]
axolotl_all = axolotl_all[,!(axolotl_all$cellclusters %in% c("GABAergic neuron", "ETV1+ neuron",
                                                             "neuron"))]
mouse_all = mouse_all[,!(mouse_all$classes %in% c("Spinal cord neurons", "Sympathetic neurons",
                                                  "Spinal cord neurons", "Peripheral sensory neurons",
                                                  "Enteric neurons", "Cerebellum neurons"))]
mouse_all = mouse_all[,!(mouse_all$subclasses %in% c("Schwann cells", "Olfactory ensheathing cells",
                                                     "Olfactory inhibitory neurons", "Enteric glia"))]
drerio_all = drerio_all[,drerio_all$classes!="doublets"]
bfinch_all = bfinch_all[,bfinch_all$classes!="Erythrocytes"]
```

Put Seurats in list

```{r}
srat_list = list("paintedturtle" = turtle_all, "axolotl" = axolotl_all, "lizard" = lizard_all,
                 "mouse" = mouse_all, "human" = human_all, "zebrafinch" = zfinch_all, 
                 "bengalesefinch" = bfinch_all, "zebrafish" = drerio_all)
```

Save Seurat list

```{r}
saveRDS(srat_list, file = "data/processed/all_sp_sratList.RDS")
```



# Ortholog matching
Load necessary orthology data

```{r}
ort_alt_ens_list = readRDS(file = "data/eggNOG/ort_alt_ens_list.RDS")
ort_alt_egg_list = readRDS(file = "data/eggNOG/ort_alt_egg_list.RDS")

ort_list_use = c(ort_alt_ens_list, 
                 ort_alt_egg_list[which(grepl("axolotl", names(ort_alt_egg_list)) |
                                          grepl("lizard", names(ort_alt_egg_list)))])
bfmats = ort_alt_egg_list[which(grepl("zebrafinch", names(ort_alt_egg_list)))]
names(bfmats) = gsub("zebrafinch", "bengalesefinch", names(bfmats))
ort_list_use = c(ort_list_use, bfmats)

# because zebrafinch and bengalese finch are using the same reference, 
## we're doing a mock matching table using zebrafinch
allzfinch = lapply(ort_list_use[grep("zebrafinch", names(ort_list_use))], 
                   function(x) unique(x[,grepl("zebrafinch", colnames(x))]))
allzfinch = unique(Reduce(rbind, allzfinch))
mockzfinch = cbind(allzfinch, allzfinch, c("ortholog_one2one"))
colnames(mockzfinch)[3:5] = c("bengalesefinchgeneid", "bengalesefinchgenename", "homology.type")
mockbfinch = cbind(allzfinch, allzfinch, c("ortholog_one2one"))
colnames(mockbfinch)[c(1,2,5)] = c("bengalesefinchgeneid", "bengalesefinchgenename", "homology.type")
ort_list_use$zebrafinch.vs.bengalesefinch = mockzfinch
ort_list_use$bengalesefinch.vs.zebrafinch = mockbfinch
```

Function for ortholog matching

```{r}
# get orthologs for all species pairs
ortMatch = function(ort_l, sp_vec, ort_scope = "ortholog_one2one"){
  sp_p_mat = combn(sp_vec, 2)
  for(i in 1:ncol(sp_p_mat)){
    sp_pair = sp_p_mat[,i]
    ort_tab = ort_l[[paste0(sp_pair, collapse = ".vs.")]]
    if(!is.null(ort_scope)){
      ort_tab = ort_tab[ort_tab$homology.type==ort_scope,]
    }

    colnames(ort_tab) = gsub("genename", "", colnames(ort_tab))
    if(i==1){
      ort_all = unique(ort_tab[,c(2,4)])
    } else{
      ort_all = unique(merge(ort_all, unique(ort_tab[,c(2,4)]), 
                             by = intersect(colnames(ort_all), colnames(ort_tab))))
    }
  }
  
  return(ort_all)
}
```

Function for subsetting Seurats

```{r}
seuratOrthologs = function(s_l, o_all, min_cnts = 50){
  # only orthologs present in the Seurat objects
  for(sp in names(s_l)){
    o_all = o_all[o_all[,sp] %in% rownames(s_l[[sp]]),]
  }
  
  # get a unique name for the genes
  allg = unlist(o_all)
  jointnames = c()
  rep_n = list()
  for(i in 1:nrow(o_all)){
    nname = unlist(lapply(o_all[i,], function(x) sum(allg==x)))
    g = unlist(o_all[i,names(nname[nname==1])])
    # if there are acceptable unique names
    if(any(nname==1) & (!is.null(g) & !all(grepl("..", g, fixed = T)))){ # no weird axololt annot
      g = g[!grepl("..", g, fixed = T)] # no weird axololt annot
      g = c(g[!grepl("LOC", g)], g[grepl("LOC", g)])[1] #prioritize non-LOC
      jointnames = c(jointnames, g)
      rep_n[[g]] = 1
    } else{ # we'll add a number to those names
      g = unlist(o_all[i,])
      g = g[!grepl("..", g, fixed = T)] # no weird axololt annot
      g = c(g[!grepl("LOC", g)], g[grepl("LOC", g)])[1] #prioritize non-LOC
      if(g %in% names(rep_n)){
        gn = rep_n[[g]]
        rep_n[[g]] = gn+1
        jointnames = c(jointnames, paste0(g, "-", gn+1))
      } else{
        rep_n[[g]] = 1
        jointnames = c(jointnames, paste0(g, "-", 1))
      }
    }
  }
  
  # filter the Seurats by the orthologs
  ## this requires redoing the Seurat with new names, since some genes will be duplicated
  for(sp in names(s_l)){
    cnts = s_l[[sp]]@assays$RNA@counts[o_all[,sp],]
    rownames(cnts) = jointnames
    m = s_l[[sp]]@meta.data
    
    # filter out cells with fewer than min_cnts
    keep = colnames(cnts)[colSums(cnts)>=min_cnts]
    
    s_l[[sp]] = CreateSeuratObject(cnts[,keep], project = sp, meta.data = m[keep,])
  }
  
  return(s_l)
}
```

Get ortholog-matched Seurats

```{r}
srat_ort = seuratOrthologs(srat_list, 
                           ortMatch(ort_list_use, names(srat_list), 
                                    ort_scope = NULL))
lapply(srat_ort, dim)

srat_ort1 = seuratOrthologs(srat_list, 
                            ortMatch(ort_list_use, names(srat_list), 
                                     ort_scope = "ortholog_one2one"))
lapply(srat_ort1, dim)
```



# Integration
Normalise data

```{r}
for(sp in names(srat_ort)){
  print(sp)
  srat_ort[[sp]] = suppressWarnings(SCTransform(srat_ort[[sp]], do.correct.umi = T, verbose = F,
                                                seed.use = 1, vars.to.regress = "nCount_RNA",
                                                variable.features.rv.th = 1, return.only.var.genes = F,
                                                variable.features.n = NULL))
  srat_ort1[[sp]] = suppressWarnings(SCTransform(srat_ort1[[sp]], do.correct.umi = T, verbose = F,
                                                 seed.use = 1, vars.to.regress = "nCount_RNA",
                                                 variable.features.rv.th = 1, return.only.var.genes = F,
                                                 variable.features.n = NULL))
}
```

Seurats with all merged

```{r}
srat_ort_all = Reduce(merge, srat_ort)
srat_ort_all = AddMetaData(srat_ort_all, 
                           metadata = unlist(lapply(names(srat_ort), 
                                                    function(x) rep(x, ncol(srat_ort[[x]])))),
                           col.name = "species")

srat_ort1_all = Reduce(merge, srat_ort1)
srat_ort1_all = AddMetaData(srat_ort1_all, 
                           metadata = unlist(lapply(names(srat_ort1), 
                                                    function(x) rep(x, ncol(srat_ort1[[x]])))),
                           col.name = "species")
```


## Seurat rPCA
Integration with all ortholog

```{r}
feat_list = list("all" = rownames(srat_ort_all@assays$SCT@scale.data), 
                 "allvar" = unique(unlist(lapply(srat_ort, VariableFeatures))), 
                 "commonvar" = Reduce(intersect, lapply(srat_ort, VariableFeatures)))
varindat = lapply(srat_ort, function(x) feat_list$allvar %in% rownames(x))
feat_list$allvar = feat_list$allvar[rowSums(Reduce(cbind, varindat))==length(srat_ort)]
srat_rpca_all_l = list()
for(n in names(feat_list)){
  print(n)
  integr_feat = feat_list[[n]]
  
  # calculate possible missing Pearson residuals for SCTransform
  srat_ort_int = PrepSCTIntegration(srat_ort, anchor.features = integr_feat, verbose = T)
  
  srat_ort_int <- lapply(X = srat_ort_int, FUN = function(x) {
    x <- ScaleData(x, features = integr_feat, verbose = FALSE)
    x <- RunPCA(x, features = integr_feat, verbose = FALSE)
  })
  
  # finding the anchors for integration
  maxpc = if(n=="commonvar") 28 else 50
  all_cell_anchors = FindIntegrationAnchors(srat_ort_int, normalization.method = "SCT", 
                                            anchor.features = integr_feat, reduction = "rpca",
                                            assay = rep("SCT", length(srat_ort_int)), 
                                            dims = 1:maxpc, verbose = T)
  
  # actual data integration
  allgenes = rownames(srat_ort_int[[1]]@assays$RNA@counts)
  all_cell_integr = IntegrateData(all_cell_anchors, normalization.method = "SCT", dims = 1:maxpc,
                                  verbose = T, features.to.integrate = allgenes)
  
  # run PCA and UMAP to see how it looks
  maxpc = if(n=="commonvar") 28 else 30
  all_cell_integr = RunPCA(all_cell_integr, verbose = F)
  all_cell_integr = RunUMAP(all_cell_integr, dims = 1:maxpc)
  
  all_cell_integr = AddMetaData(all_cell_integr, 
                             metadata = unlist(lapply(names(srat_ort), 
                                                      function(x) rep(x, ncol(srat_ort[[x]])))),
                             col.name = "species")
  srat_rpca_all_l[[n]] = all_cell_integr
}
```

Integration with one2one ortholog

```{r}
feat_list = list("all" = rownames(srat_ort1_all@assays$SCT@scale.data), 
                 "allvar" = unique(unlist(lapply(srat_ort1, VariableFeatures))), 
                 "commonvar" = Reduce(intersect, lapply(srat_ort1, VariableFeatures)))
varindat = lapply(srat_ort1, function(x) feat_list$allvar %in% rownames(x))
feat_list$allvar = feat_list$allvar[rowSums(Reduce(cbind, varindat))==length(srat_ort1)]
srat_rpca_one_l = list()
for(n in names(feat_list)[1]){
  print(n)
  integr_feat = feat_list[[n]]
  
  # calculate possible missing Pearson residuals for SCTransform
  srat_ort1_int = PrepSCTIntegration(srat_ort1, anchor.features = integr_feat, verbose = T)
  
  srat_ort1_int <- lapply(X = srat_ort1_int, FUN = function(x) {
    x <- ScaleData(x, features = integr_feat, verbose = FALSE)
    x <- RunPCA(x, features = integr_feat, verbose = FALSE)
  })
  
  # finding the anchors for integration
  all_cell_anchors = FindIntegrationAnchors(srat_ort1_int, normalization.method = "SCT", 
                                            anchor.features = integr_feat, reduction = "rpca",
                                            assay = rep("SCT", length(srat_ort1_int)), 
                                            dims = 1:30, verbose = T)
  
  # actual data integration
  allgenes = rownames(srat_ort1_int[[1]]@assays$RNA@counts)
  all_cell_integr = IntegrateData(all_cell_anchors, normalization.method = "SCT", dims = 1:50,
                                  verbose = T, features.to.integrate = allgenes)
  
  # run PCA and UMAP to see how it looks
  all_cell_integr = RunPCA(all_cell_integr, verbose = F)
  all_cell_integr = RunUMAP(all_cell_integr, dims = 1:30)
  
  all_cell_integr = AddMetaData(all_cell_integr, 
                             metadata = unlist(lapply(names(srat_ort1), 
                                                      function(x) rep(x, ncol(srat_ort1[[x]])))),
                             col.name = "species")
  srat_rpca_one_l[[n]] = all_cell_integr
}
```


## Combined
rPCA + Harmony, all ortholog

```{r}
srat_rpca_harm_all_l = list()
for(n in names(srat_rpca_all_l)){
  print(n)
  # Run Harmony
  cell_integr = RunHarmony(srat_rpca_all_l[[n]], "species", tau = 30, 
                           plot_convergence = F, assay.use = "SCT")
  
  # Run UMAP on Harmony dimensions
  cell_integr = RunUMAP(cell_integr, reduction = "harmony", dims = 1:10)
  
  srat_rpca_harm_all_l[[n]] = cell_integr
}
```

rPCA + Harmony, one2one ortholog

```{r}
srat_rpca_harm_one_l = list()
for(n in names(srat_rpca_one_l)){
  print(n)
  # Run Harmony
  cell_integr = RunHarmony(srat_rpca_one_l[[n]], "species", tau = 30, 
                           plot_convergence = F, assay.use = "SCT")
  
  # Run UMAP on Harmony dimensions
  cell_integr = RunUMAP(cell_integr, reduction = "harmony", dims = 1:10)
  
  srat_rpca_harm_one_l[[n]] = cell_integr
}
```

Plots for all genes

```{r, fig.height=13, fig.width=13}
for(n in names(srat_rpca_all_l)){
  plt11 = DimPlot(srat_rpca_all_l[[n]], reduction = "umap", group.by = "species", shuffle = T)+
    labs(title = paste0("rPCA, all orthologs, ", n))+
    theme(aspect.ratio = 1)
  plt12 = DimPlot(srat_rpca_all_l[[n]], reduction = "umap", group.by = "classes", shuffle = T)+
    labs(title = paste0("rPCA, all orthologs, ", n))+
    theme(aspect.ratio = 1)
  print(plt11/plt12)
  
  plt41 = DimPlot(srat_rpca_harm_all_l[[n]], reduction = "umap", group.by = "species", shuffle = T)+
    labs(title = paste0("rPCA+Harmony, all orthologs, ", n))+
    theme(aspect.ratio = 1)
  plt42 = DimPlot(srat_rpca_harm_all_l[[n]], reduction = "umap", group.by = "classes", shuffle = T)+
    labs(title = paste0("rPCA+Harmony, all orthologs, ", n))+
    theme(aspect.ratio = 1)
  print(plt41/plt42)
}
```

Plots for one2one genes

```{r, fig.height=13, fig.width=13}
for(n in names(srat_rpca_one_l)){
  plt11 = DimPlot(srat_rpca_one_l[[n]], reduction = "umap", group.by = "species", shuffle = T)+
    labs(title = paste0("rPCA, one2one orthologs, ", n))+
    theme(aspect.ratio = 1)
  plt12 = DimPlot(srat_rpca_one_l[[n]], reduction = "umap", group.by = "classes", shuffle = T)+
    labs(title = paste0("rPCA, one2one orthologs, ", n))+
    theme(aspect.ratio = 1)
  print(plt11/plt12)
  
  plt41 = DimPlot(srat_rpca_harm_one_l[[n]], reduction = "umap", group.by = "species", shuffle = T)+
    labs(title = paste0("rPCA+Harmony, one2one orthologs, ", n))+
    theme(aspect.ratio = 1)
  plt42 = DimPlot(srat_rpca_harm_one_l[[n]], reduction = "umap", group.by = "classes", shuffle = T)+
    labs(title = paste0("rPCA+Harmony, one2one orthologs, ", n))+
    theme(aspect.ratio = 1)
  print(plt41/plt42)
}
```




